### PR TITLE
#962 Msg API - switch to implicit conversion

### PR DESCRIPTION
**Pull request by pnstickne**
Changes theMsg()'s sendMsg and broadcastMsg to perform an implicit std::move on the `MsgPtr&` argument.

Change all core usages to supply a `MsgPtr&` (and all but two usage in tests).

This change can make some code that worked before fail is the implicit move-like semantics are not honored.

Notes:

This currently uses an implicit proxy to unify both `MsgPtr&` and `Msg*`. However, this type can be eliminated if dropping `Msg*` directly and always accepting `MsgPtr&`.

Over all the current code in VT, the pattern of creating a message is "the method" so there is no need to accept a `MsgPtr` or `MsgPtr&&`.

The implicit type is still very hand for transitioning over.

### CODE REVIEW

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by PhilMiller** [comment]
Looks good so far. Looking forward to seeing all of the `msg.get()` in caller code disappear

File: src/vt/messaging/active.h [hunk 1]
```diff
@@ -178,6 +178,35 @@ struct BufferedActiveMsg {
   { }
 };
 
+/**
```
Comment (lifflander) [inline src/vt/messaging/active.h:181]: Can we put this in its own header file? I think we will want to use if in other components without including active.h

File: src/vt/registry/auto/auto_registry.h [hunk 2]
```diff
@@ -73,6 +73,9 @@ HandlerType makeAutoHandlerObjGroup(HandlerControlType control);
 template <typename MsgT, ActiveTypedFnType<MsgT>* f>
 HandlerType makeAutoHandler(MsgT* const msg);
```
Comment (lifflander) [inline src/vt/registry/auto/auto_registry.h:74]: We should delete the `msg` argument. It's not used anywhere

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tutorial/tutorial_1g.h  -3
+ tests/unit/pipe/test_callback_send.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tutorial/tutorial_1g.h  -3
+ tests/unit/pipe/test_callback_send.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tutorial/tutorial_1g.h  -3
+ tests/unit/pipe/test_callback_send.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by pnstickne** [comment]
@lifflander "Works Here", ahah ^_^

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tutorial/tutorial_1g.h  -3
+ tests/unit/pipe/test_callback_send.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

File: src/vt/collective/reduce/reduce.impl.h [hunk 3]
```diff
@@ -238,7 +238,8 @@ void Reduce::startReduce(detail::ReduceStamp id, bool use_num_contrib) {
           scope_.str(), detail::stringizeStamp(id), root, this_node
         );
 
-        theMsg()->sendMsg<MsgT,ReduceManager::reduceRootRecv<MsgT>>(root,typed_msg);
+        auto pmsg = promoteMsg(typed_msg);
+        theMsg()->sendMsg<MsgT,ReduceManager::reduceRootRecv<MsgT>>(root, pmsg);
```
Comment (PhilMiller) [inline src/vt/collective/reduce/reduce.impl.h:242]: Should we be able to eliminate the `typed_msg` above and `pmsg` here, and just pass `msg` directly to this call and the one below?

File: src/vt/messaging/message/smart_ptr.h [hunk 4]
```diff
@@ -226,6 +232,40 @@ struct MsgSharedPtr final {
   MsgPtrImplBase* impl_ = nullptr;
 };
 
+/**
+ * \brief Helper to unify 'stealing' message ownership.
+ *
+ * This type is not intented to be used directly. It uses implicit conversion
+ * construtors to perform a 'std::move' on a \c MsgPtr<MsgT>&.
```
Comment (PhilMiller) [inline src/vt/messaging/message/smart_ptr.h:239]: Typo 'constructors'

File: src/vt/vrt/collection/balance/node_stats.cc [hunk 5]
```diff
@@ -156,8 +156,10 @@ ElementIDType NodeStats::getNextElm() {
 void NodeStats::releaseLB() {
   using MsgType = CollectionPhaseMsg;
   auto msg = makeMessage<MsgType>();
-  theMsg()->broadcastMsg<MsgType,CollectionManager::releaseLBPhase>(msg.get());
-  CollectionManager::releaseLBPhase(msg.get());
+  auto msg_hold = promoteMsg(msg.get());
+  theMsg()->broadcastMsg<MsgType,CollectionManager::releaseLBPhase>(msg);
+
+  CollectionManager::releaseLBPhase(msg_hold.get());
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/node_stats.cc:162]: This repeated pattern of get, promote, broadcast, and deliver further favors making broadcast include local delivery.

File: src/vt/termination/termination.cc [hunk 6]
```diff
@@ -439,12 +439,13 @@ bool TerminationDetector::propagateEpoch(TermStateType& state) {
...
       theMsg()->markAsTermMessage(msg);
-      theMsg()->sendMsg<TermCounterMsg, propagateEpochHandler>(parent, msg.get());
+      auto msg_hold = promoteMsg(msg.get()); // keep after send
+      theMsg()->sendMsg<TermCounterMsg, propagateEpochHandler>(parent, msg);
 
       vt_debug_print_verbose(
         term, node,
         "propagateEpoch: sending to parent: {}, msg={}, epoch={:x}, wave={}\n",
-        parent, print_ptr(msg.get()), state.getEpoch(), state.getCurWave()
+        parent, print_ptr(msg_hold.get()), state.getEpoch(), state.getCurWave()
```
Comment (PhilMiller) [inline src/vt/termination/termination.cc:448]: This print can move before the send, to avoid needing to hold onto a copy of the message pointer

**Comment by PhilMiller** [comment]
This is good progress. I like that this makes `promoteMsg` an indicator that something weird is being done, and that the surrounding code needs more care and attention on it. Ideally, we'd be able to reduce the number of call sites accordingly, now that we can see them clearly.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tests/unit/pipe/test_callback_send.cc  -2
+ tutorial/tutorial_1g.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tests/unit/pipe/test_callback_send.cc  -2
+ tutorial/tutorial_1g.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

File: src/vt/group/collective/group_info_collective.cc [hunk 7]
```diff
@@ -388,12 +389,13 @@ void InfoColl::upTree() {
...
       group,op,is_in_group,static_cast<NodeType>(subtree_),child,0,extra
     );
-    theMsg()->sendMsg<GroupCollectiveMsg,upHan>(p, msg.get());
+    theMsg()->sendMsg<GroupCollectiveMsg,upHan>(p, msg);
     /*
      *  Forward all the children messages up the tree (up to 2 of them)
      */
     for (std::size_t i = 0; i < msg_in_group.size(); i++) {
-      theMsg()->sendMsg<GroupCollectiveMsg,upHan>(p, msg_in_group[i].get());
+      auto msg_out = promoteMsg(msg_in_group[i].get()); // #692 - promote/steal
```
Comment (PhilMiller) [inline src/vt/group/collective/group_info_collective.cc:397]: Typo here of 962

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tutorial/tutorial_1g.h  -3
+ tests/unit/pipe/test_callback_send.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

File: src/vt/messaging/active.h (hunk 1 above)
Comment (pnstickne) [reply src/vt/messaging/active.h]: It's now in with MsgPtr, as they are very much .. dependent?

File: src/vt/vrt/collection/balance/node_stats.cc (hunk 5 above)
Comment (pnstickne) [reply src/vt/vrt/collection/balance/node_stats.cc:162]: @PhilMiller `broadcastMsgIncludingSelf`? A mouthful..

File: src/vt/termination/termination.cc (hunk 6 above)
Comment (pnstickne) [reply src/vt/termination/termination.cc:448]: Done.

File: src/vt/group/collective/group_info_collective.cc (hunk 7 above)
Comment (pnstickne) [reply src/vt/group/collective/group_info_collective.cc:397]: Improved/fixed.

File: src/vt/collective/reduce/reduce.impl.h (hunk 3 above)
Comment (pnstickne) [reply src/vt/collective/reduce/reduce.impl.h:242]: Yes. Done.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tutorial/tutorial_1g.h  -3
+ tests/unit/memory/test_memory_lifetime.cc  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/registry/auto/auto_registry_impl.h  2
         

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -4
+ tutorial/tutorial_1g.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

File: src/vt/messaging/message/smart_ptr.h (hunk 4 above)
Comment (pnstickne) [reply src/vt/messaging/message/smart_ptr.h:272]: Fixed.

File: src/vt/registry/auto/auto_registry.h (hunk 2 above)
Comment (pnstickne) [reply src/vt/registry/auto/auto_registry.h:74]: Cleaned up the auto-handlers and remove passing messages (or usually nullptr).

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -4
+ tutorial/tutorial_1g.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by pnstickne** [comment]
The failing test is addressed in #990. After #990 is merged the MsgPtrThief tests here should be updated to be no-MPI.

File: src/vt/messaging/message/smart_ptr.h (hunk 4 above)
Comment (PhilMiller) [reply src/vt/messaging/message/smart_ptr.h:272]: ... It still reads 'construtors' (missing 'c'). Neglected push?

**Comment by bradybray** [comment]
Looks good save for the one failed build.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -4
+ tutorial/tutorial_1g.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/memory/test_memory_lifetime.cc  -4
+ tutorial/tutorial_1g.h  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5958715&bid=19534286)

**Review by lifflander — APPROVED** [review]

**Pull request merged**